src/core/qfieldxmlhttprequest.h

```diff
@@ -0,0 +1,314 @@
+    void setOnerror( const QJSValue &cb ) { mOnError = cb; }
+
+    QJSValue onload() const { return mOnLoad; }
+    void setOnload( const QJSValue &cb ) { mOnLoad = cb; }
+
+    QJSValue onloadend() const { return mOnLoadEnd; }
+    void setOnloadend( const QJSValue &cb ) { mOnLoadEnd = cb; }
+
+  signals:
+    void readyStateChanged();
+    void responseChanged();
+    void timeoutChanged();
+    void autoDeleteChanged();
+
+  private:
+    // Returns true if open() was called with valid method+URL.
+    bool isOpen() const;
+
+    // Transitions readyState and emits onreadystatechange.
+    void setReadyState( ReadyState state );
+
+    // Clears response-related members (status/text/body/type/url).
+    void clearResponse();
+
+    // Resets per-request flags (finalized/aborted/timedOut) and clears response.
+    void resetForNewRequest();
+
+    /**
+   * \brief Disconnects signals and disposes current reply objects.
+   * \param abortNetwork If true, aborts the network operation before deleting the reply.
+   *    * This method is used both for resend and destruction. It must never call JS callbacks.
+   */
+    void cleanupReply( bool abortNetwork );
+
+    // Creates and starts the underlying NetworkReply based on method/body.
+    void startRequest( const QVariant &body );
+
+    // Hooks high-level NetworkReply signals (progress/errors/finished/raw reply changes).
+    void hookReply( NetworkReply *reply );
+
+    // Hooks current raw QNetworkReply to derive readyState transitions.
+    void hookRawReply( QNetworkReply *rawReply );
+
+    // Finalizes response members from raw reply data.
+    void finalizeFromRawReply();
+
+    // Finalizes as an error message payload (stored in responseText/response).
+    void finalizeAsError( const QString &msg, bool emitResponse = true );
+
+    // Serializes body into raw bytes (with optional inferred Content-Type).
+    QByteArray bodyToBytes( const QVariant &body, QString *outContentType = nullptr ) const;
+
+    // Detects explicit "file://..." strings to decide multipart behavior.
+    bool bodyContainsFileUrls( const QVariant &body ) const;
+
+    // Builds multipart/form-data body; returns nullptr on invalid structure.
+    QHttpMultiPart *buildMultipart( const QVariant &body ) const;
+
+    // Returns true if a local file path is allowed to be uploaded.
+    bool isAllowedLocalUploadPath( const QString &localPath ) const;
+
+    // Calls a QML callback safely
+    void call( const QJSValue &cb, const QJSValueList &args = {} );
+
+    // Basic newline detection helper for header protection.
+    static bool containsNewlines( const QString &s );
+
+    // Collect final response headers from raw reply
+    void collectResponseHeaders( QNetworkReply *rawReply );
+    static QString normalizeHeaderName( const QString &s );
+
+    // Schedules deleteLater() if autoDelete is enabled
+    void maybeAutoDelete();
+
+  private:
+    QPointer<NetworkReply> mReply;
```

Comment:
Don't use the NetworkReply here, simply use QNetworkReply.

That means you can remove the hookNetworkReply too.

NetworkReply is specifically used for network communication between QField and QFieldCloud , we don't need them nor want them exposed here to plugin authors etc.